### PR TITLE
[stdlib] Add the new `Hasher` trait with some unit tests

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
This was done in cooperation with @mzaks  we want to introduce the new hasher progressively to make it easier to review and flush out the compiler bugs.

New names will be used to avoid conflicts during the transition.

The source PR is this one: https://github.com/modularml/mojo/pull/2619

EDIT: Co-authored-by doesn't seem to work correctly, strange...
EDIT 2: That Co-authored thing really doesn't want to work. Would it be possible to add `Co-authored-by: Maxim Zaks <maxim.zaks@gmail.com>` internally when merging the PR? Thank you very much.

### CODE REVIEW

**Comment by JoeLoser** [comment]
!sync

**Comment by JoeLoser** [comment]
!sync

File: stdlib/src/builtin/_hasher.mojo [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+# limitations under the License.
+# ===----------------------------------------------------------------------=== #
+
+
+trait _HashableWithHasher:
+    fn __hash__[H: _Hasher](self, inout hasher: H):
+        ...
+
+
+trait _Hasher:
```
Comment (JoeLoser) [inline stdlib/src/builtin/_hasher.mojo:20]: **Question** Are we knowingly omitting the `_update_with_bytes` API from https://github.com/modularml/mojo/blob/32e7c06bd832f9e781d00bf11cc6911c7b678c18/proposals/improved-hash-module.md#introduce-a-hasher-trait for now?

**Review by JoeLoser — APPROVED** [review]
Looks good, thank for splitting this up so we can iterate on the pieces in the midst of compiler bugs/avoiding massive PRs.  It will be easier to land and migrate for everyone.

Just had one open question, but otherwise, happy to merge it.  Thank you!

File: stdlib/src/builtin/_hasher.mojo (hunk 1 above)
Comment (gabrieldemarmiesse) [reply stdlib/src/builtin/_hasher.mojo:20]: Yeah we'll add it in a future PR, that is intentional

Comment (JoeLoser) [reply stdlib/src/builtin/_hasher.mojo:20]: SG, I'll land this now internally. Thanks!

**Pull request closed without merging**